### DISCUSSION TITLE
How did you setup eslint and prettier for large typescript project?

### DISCUSSION THREAD

**Discussion by nyngwang** [stevearc/qf_helper.nvim#7] [Q&A] [answered]
I need your advice about the best practice(currently) to setup both `eslint` and `prettier` for large typescript projects. I will first describe why I favor one approach over the other, then describe my setup for that approach, finally some questions regarding my setup.

__Why `nvim-lsp-config`__: The priority of all considerations in my mind is of order: `stability > setup-complexity > memory-consumption`(It seems that `eslint_d` would consume more than 1G memory). I read some posts saying that `null-ls` is a good choice, while some people are complaining that the setup is hard. Then I read another post saying that now we have `ESLint` language server. Since I'm more familiar with `nvim-lsp-config`(which can install `ESLint`-language server easily), I adopt the last approach today.

__Here is my setup__:

- I use `nvim-lsp-installer` to install the `eslint` language server.
- on server ready, I passed the following `M` as the `config` of `...eslint.setup(config)`:

    ```lua
    M = {
      capabilities = require('cmp_nvim_lsp').update_capabilities(vim.lsp.protocol.make_client_capabilities()),
      autostart = true,
      on_attach = function (...)
        require('lsp_config.common_mappings')(...) -- some common mappings from another file
        -- this `autocmd` is the recommended setup from the nvim-lspconfig wiki
        vim.cmd([[
          augroup eslint_fixall_on_save
            autocmd!
            autocmd BufWritePre <buffer> lua vim.cmd('EslintFixAll')
          augroup END
        ]])
      end,
      settings = {
        packageManager = 'yarn', -- the project I'm working on use `yarn` instead of `npm`
      }
    }
    
    return M
    ```

__Some questions__ I still have regarding this `M`:

- Is this setup enough for what I want to achieve? I want to be able to use a command to fix all eslint error, and automate this on save when possible.
- The project I'm working on install all packages via `yarn` with a custom/private source. Am I correct to set the `packageManager = 'yarn'`?
- Some comment says that `%!npx prettier` is enough. Is this a good idea?

---

For `prettier`, since I just saw that you're one of the contributors of [neoformat](https://github.com/sbdchd/neoformat), I decided to use it. Would I have to do something (e.g. correctly set the `pwd` before run) to make sure that the project `.prettierrc` is referenced?

Thanks for your reading! If this is not appropriate to be asked here then I will delete it...

**Comment by stevearc** [comment, marked as answer]
Yeah, these questions are about a lot of things that I don't use. I would recommend posting in www.reddit.com/r/neovim or [the neovim discource](https://neovim.discourse.group/) or possibly [the matrix channel](https://app.element.io/#/room/#neovim:matrix.org).

FWIW I use null-ls now for all my autoformatting and linting. Would definitely recommend it over neoformat, but not sure how it stacks up against the ESLint language server.

**Reply by nyngwang** [reply]
I painfully resolved this today. I can share it with you later(on work now :( ). I'm using null-ls and ESLint lsp server.
